### PR TITLE
CodeGenPrepare: Add support for llvm.threadlocal.address address-mode sinking

### LINKED ISSUE #87437: llvm.threadlocal.address changes regressed code using TLS
Writing this mostly as a FYI and wondering if others have observed this or are looking into this too?

We are in the process of upgrading compiler versions and noticed a 0.5% regression in HHVM which is a heavy user of TLS. This was bisected to the changes in https://reviews.llvm.org/D125291 and https://reviews.llvm.org/D129833 that introduce `llvm.theadlocal.address`.

While I am still in the process of completely classifying (and maybe fixing) things, I believe this shows up as increased CSEing of TLS base addresses and consequently increased spilling activity. We also seem to be unable to rematerialize or fold such accesses further increasing register pressure and hence spills.

Example of bad codegen pattern (within a bigger function):
```
        movq    %fs:0x0, %rcx
        leaq    -0x2540(%rcx), %rcx
        movq    %rcx, -0x68(%rbp)
        movq    %fs:-0x2540, %rcx
        ...
        movq    -0x68(%rbp), %rcx
        movq    (%rcx), %rcx
```

### PR DESCRIPTION
**Pull request by MatzeB**
Depending on the TLSMode many thread-local accesses on x86 can be expressed by adding a %fs: segment register to an addressing mode. Even if there are mutliple users of a `llvm.threadlocal.address` intrinsic it is generally not worth sharing the value in a register but instead fold the %fs access into multiple addressing modes.

Hence this changes CodeGenPrepare to duplicate the `llvm.threadlocal.address` intrinsic as necessary.

Introduces a new `TargetLowering::addressingModeSupportsTLS` callback that allows targets to indicate whether TLS accesses can be part of an addressing mode.

This is fixing a performance problem, as this folding of TLS-accesses into multiple addressing modes happened naturally before the introduction of the `llvm.threadlocal.address` intrinsic, but regressed due to `SelectionDAG` keeping things in registers when accessed across basic blocks, so CodeGenPrepare needs to duplicate to mitigate this. We see a ~0.5% recovery in a codebase with heavy TLS usage (HHVM).

This fixes #87437

### CODE REVIEW

**Comment by MatzeB** [comment]
I also submitted #87841 to verify the assumption that arg0 of `threadlocal.address` is a thread-local `GlobalVariable`. (Turns out it currently fails 1 test, but it's debatable whether the test is okay).

File: llvm/test/CodeGen/X86/tls-multi-use.ll [hunk 1]
```diff
@@ -0,0 +1,149 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -o - %s | FileCheck %s
+; RUN: llc -o - -relocation-model=pic %s | FileCheck --check-prefix=CHECK_PIC %s
+target triple = "x86_64--linux-gnu"
+
+@foo = dso_local thread_local global i32 0, align 4
+@foo_nonlocal = thread_local global i32 0, align 4
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/tls-multi-use.ll:7]: Not used?

**Review by ChuanqiXu9 — APPROVED** [review]
I am not familiar with the codes in the backend. But it looks good in the high level. Please wait a few days in case there are more comments.

File: llvm/test/CodeGen/X86/tls-multi-use.ll (hunk 1 above)
Comment (MatzeB) [reply llvm/test/CodeGen/X86/tls-multi-use.ll]: good catch, this was intended for the 2nd function so we see `dso_local` and non-`dso_local` variables referenced in the test.

**Comment by jyknight** [comment]
Hm. I think this is an unfortunate way to fix the issue. Long-term, I'd like to get to the point where it is no longer valid to reference a TLS value directly in IR as if it was a simple address, but instead, require that accesses MUST always go through a call. We aren't in that ideal state _now_, but this goes further away from it instead of further towards it.

**Comment by MatzeB** [comment]
> Hm. I think this is an unfortunate way to fix the issue. Long-term, I'd like to get to the point where it is no longer valid to reference a TLS value directly in IR as if it was a simple address, but instead, require that accesses MUST always go through a call. We aren't in that ideal state now, but this goes further away from it instead of further towards it.

This isn't really changing how IR behaves, at least it's not changing anything in the actual transform/optimization phases of LLVM...

I was thinking of this as just yet another hack/workaround where massaging the IR in `CodeGenPrepare` is just way easier than  doing things in SelectionDAG because of its "basic block at a time" design. From all I can tell it is somewhat fundamentally baked into SelectionDAG today that instructions read across basic blocks need registers assigned and only `Value`s not tied to instructions can be seen immediately. I initially tried fixing this in SelectionDAG but that would need more fundamental changes there than seemed reasonable for a small improvement like this...

**Comment by nikic** [comment]
I agree with @jyknight that we should avoid dropping the threadlocal.address intrinsics in this way. The long-term plan for TLS was to change TLS globals to return a token value, and make threadlocal.address accept that token and return the actual pointer. This change wouldn't be compatible with that, as we'd replace the pointer with the token.

You do need to address this in CGP, but I'd expect it to be in the same way as we do it for address calculations in general: By duplicating/sinking. Possibly this is something that should be directly integrated in the address mode logic (as this is ultimately about address modes), but if nothing else you can duplicate the threadlocal.address intrinsics separately.

**Comment by MatzeB** [comment]
Reworked this change in terms of `CodeGenPrepare` address-mode sinking logic and enabled the duplication on `llvm.threadlocal.address` for X86 `TLSModel::LocalExec` and `TLSModel::InitialExec` where it improves code quality.

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 2]
```diff
@@ -5082,6 +5082,16 @@ bool AddressingModeMatcher::matchOperationAddr(User *AddrInst, unsigned Opcode,
     }
     return true;
   }
+  case Instruction::Call:
+    if (IntrinsicInst *II = dyn_cast<IntrinsicInst>(AddrInst)) {
+      if (II->getIntrinsicID() == Intrinsic::threadlocal_address) {
+        GlobalValue &GV = cast<GlobalValue>(*II->getArgOperand(0));
+        if (TLI.addressingModeSupportsTLS(GV)) {
```
Comment (nikic) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:5089]: No braces on single-line if.

File: llvm/test/CodeGen/X86/codegen-prepare-addrmode-tls.ll [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+; RUN: llc -o - %s -relocation-model=pic -enable-tlsdesc | FileCheck %s --check-prefix=TLSDESC
+
+target triple = "x86_64--linux-gnu"
+
+declare void @effect()
+declare nonnull ptr @llvm.threadlocal.address.p0(ptr nonnull)
+
+@foo_local = dso_local thread_local(localexec) global i32 0, align 4
+
+define i32 @func_local_tls(i32 %arg0, i32 %arg1) {
```
Comment (nikic) [inline llvm/test/CodeGen/X86/codegen-prepare-addrmode-tls.ll:13]: Add nounwind to avoid the cfi comments.

**Review by nikic — APPROVED** [review]
LGTM, but please give it a day in case there's more feedback.

**Pull request merged**